TeamCode/src/main/java/org/firstinspires/ftc/teamcode/robot/subsystems/XDrive.java

```diff
@@ -0,0 +1,267 @@
+        front_right.setPower(fR);
+    }
+
+    public void resetAngle(SmartGamepad driver){
+        if(driver.isButtonA()){
+            reset_angle = getHeading() + reset_angle;
+        }
+    }
+
+    public double getHeading(){
+        YawPitchRollAngles angles = imu.getRobotYawPitchRollAngles();
+        double heading = angles.getYaw(AngleUnit.DEGREES);
+        if(heading < -180) {
+            heading = heading + 360;
+        }
+        else if(heading > 180){
+            heading = heading - 360;
+        }
+        heading = heading - reset_angle;
+        return heading;
+    }
+
+    public static synchronized XDrive getInstance() {
+        if (instance == null) {
+            instance = new XDrive();
+        }
+        return instance;
+    }
+    public void setPower(double FLSpeed,double FRSpeed,double BLSpeed,double BRSpeed){
+        front_left.setPower(FLSpeed);
+        front_right.setPower(FRSpeed);
+        back_left.setPower(BLSpeed);
+        back_right.setPower(BRSpeed);
+    }
+
+    public void alignAtTag(Pose2d tagPosition,SmartGamepad driver){
+        double VX = -YpositionPID.calculate(AutonomousConstants.ALIGN_AT_TAG_DISTANCE,tagPosition.getY());
+        double VY = driver.getLeftStickX();
+        double VH = -HpositionPID.calculate(0,tagPosition.getHeadingRadians())*AutonomousConstants.HEADING_APRIL_CONSTANT;
+        MotorVelocityData wheelVels = getDesiredWheelVelocities(new Pose2d(VX,VY,VH));
+        telemetry.addData("VX",VX);
+        telemetry.addData("VY",VY);
+        telemetry.addData("VH",VH);
+        telemetry.addData("X position detected from tags:", currentPose.getX());
+        telemetry.addData("Y position detected from tags:", currentPose.getY());
+        telemetry.addData("Heading detected from tags:", currentPose.getHeadingDegrees());
+        telemetry.addData("Target X position: ",AutonomousConstants.ALIGN_AT_TAG_DISTANCE);
+        telemetry.addData("Target Heading: ",0.0);
+
+
+        setPower(wheelVels.velocityFrontLeft,wheelVels.velocityFrontRight, wheelVels.velocityBackLeft,wheelVels.velocityBackRight);
+    }
+    public MotorVelocityData getDesiredWheelVelocities(Pose2d botRelativeVelocity){
+        return new MotorVelocityData().updateAppliedVelocities(botRelativeVelocity).normalize();
+    }
+    public Pose2d getCurrentPose(){
+        return currentPose;
+    }
+    public void setCurrentPose(Pose2d NewPose){currentPose = NewPose;}
+
+    public  void relativeOdometryUpdate() {
+        double dtheta;
+        Pose2d robotPoseDelta = wheelToRobotVelocities();
+
+        //double dtheta = currentPose.getHeadingRadians()+(imu.getRobotYawPitchRollAngles().getYaw(AngleUnit.RADIANS)-currentPose.getHeadingRadians());
+        if (imu.getRobotYawPitchRollAngles().getYaw(AngleUnit.RADIANS)-currentPose.getHeadingRadians()>Math.PI){
+            dtheta = -2*Math.PI+imu.getRobotYawPitchRollAngles().getYaw(AngleUnit.RADIANS)-currentPose.getHeadingRadians();
```

Comment:
use ternary instead

Reply:
did that but put values on following lines in order to not make a single huge line